### PR TITLE
CORE-10598 Reconcile the members topic with the member info table

### PR DESCRIPTION
**Pull request by williamvigorr3**
Add a reconciler to reconcile the members topic with the member info table. The member info table is only used by an MGM  Virtual Node and we don't support deletion. I have added the field is deleted to the table as this is required by the reconciler code. The reconciler will run on a non-MGM Virtual Node but this won't cause any functional issues because there won't be any records with isDeleted = true, so it won't write any tombstone records.

I have updated the `suspend` and `activate` member MGM rest commands to make the serial number mandatory (the old rest endpoints are deprecated for removal in a later version). This makes the API is less prone to user error, the MGM operator will have to check the members memberInfo for the serial number before suspending the member.

Requires https://github.com/corda/corda-api/pull/1132.

Manual testing
====
1. Test of the reconciler
I changed `publishSuspensionActivationRecords` in `publishSuspensionActivationRecords` to not publish the memberInfo record. By removing this line:
```
Record(topic = MEMBER_LIST_TOPIC, key = "${mgmShortHash}-${memberShortHash}", value = memberInfo),
```
Suspending Bob:

Eventually the reconciler runs:
```
2023-08-09 14:04:59.990 [lifecycle-coordinator-1] INFO  net.corda.processors.db.internal.reconcile.db.MemberInfoReconciler {} - Reconciling record for: O=Bob, L=London, C=GB in {"x500Name": "O=MGM, L=London, C=GB", "groupId": "f43eb984-425a-4ebf-bb12-8e6f26569a76"}'s member list.
```
The MGM's members REST endpoint shows Bob as Suspended (using `curl --insecure -u admin:admin "$API_URL/members/7CC9A374ACFB?statuses=SUSPENDED"`).

3. New and Deprecated REST Commands
Deprecated REST Commands still work with no serial:
```
william.vigor@LDNM-WHD6VPFHQC corda-runtime-os % curl --insecure -u admin:admin "https://localhost:8888/api/v1/mgm/7CC9A374ACFB/suspend/" -H 'Content-Type: application/json' -d '{"x500Name": "O=Bob, L=London, C=GB"}'
william.vigor@LDNM-WHD6VPFHQC corda-runtime-os % curl --insecure -u admin:admin "https://localhost:8888/api/v1/mgm/7CC9A374ACFB/activate/" -H 'Content-Type: application/json' -d '{"x500Name": "O=Bob, L=London, C=GB"}'
```
New REST Commands don't work without serial:
```
curl --insecure -u admin:admin "https://localhost:8888/api/v5_1/mgm/7CC9A374ACFB/suspend/" -H 'Content-Type: application/json' -d '{"x500Name": "O=Bob, L=London, C=GB", "serialNumber": " "}'
{
    "title": "The serial number must be provided in the request.",
    "status": 400,
    "details": {}
}%
william.vigor@LDNM-WHD6VPFHQC corda-runtime-os % curl --insecure -u admin:admin "https://localhost:8888/api/v5_1/mgm/7CC9A374ACFB/activate/" -H 'Content-Type: application/json' -d '{"x500Name": "O=Bob, L=London, C=GB", "serialNumber": " "}'
{
    "title": "The serial number must be provided in the request.",
    "status": 400,
    "details": {}
}%
```

### CODE REVIEW

File: libs/membership/membership-datamodel/src/main/kotlin/net/corda/membership/datamodel/MemberInfoEntity.kt [hunk 1]
```diff
@@ -55,7 +55,10 @@ class MemberInfoEntity(
     var mgmContext: ByteArray,
 
     @Column(name = "serial_number", nullable = false)
-    var serialNumber: Long,
+    val serialNumber: Long,
+
+    @Column(name = "is_deleted", nullable = false, updatable = true)
```
Comment (yift-r3) [inline libs/membership/membership-datamodel/src/main/kotlin/net/corda/membership/datamodel/MemberInfoEntity.kt:60]: Should we set the default to false?

File: libs/membership/membership-datamodel/src/main/kotlin/net/corda/membership/datamodel/MemberInfoEntity.kt [hunk 2]
```diff
@@ -55,7 +55,10 @@ class MemberInfoEntity(
     var mgmContext: ByteArray,
 
     @Column(name = "serial_number", nullable = false)
-    var serialNumber: Long,
+    val serialNumber: Long,
```
Comment (yift-r3) [inline libs/membership/membership-datamodel/src/main/kotlin/net/corda/membership/datamodel/MemberInfoEntity.kt]: I think we should keep all of them as `var` - see https://github.com/corda/corda-runtime-os/pull/4373

File: components/membership/membership-persistence-service-impl/src/test/kotlin/net/corda/membership/impl/persistence/service/handler/QueryMemberInfoHandlerTest.kt [hunk 3]
```diff
@@ -140,7 +140,8 @@ class QueryMemberInfoHandlerTest {
         "sig-$otherX500Name".toByteArray(),
         "dummySignatureSpec",
         mgmContextBytes,
-        1L
+        1L,
+        isDeleted = false
```
Comment (yift-r3) [inline components/membership/membership-persistence-service-impl/src/test/kotlin/net/corda/membership/impl/persistence/service/handler/QueryMemberInfoHandlerTest.kt:144]: Do we need to change the query in `QueryMemberInfoHandler.kt` not to return a row with `is_deleted = true`?

File: processors/db-processor/src/main/kotlin/net/corda/processors/db/internal/DBProcessorImpl.kt [hunk 4]
```diff
@@ -114,6 +117,12 @@ class DBProcessorImpl @Activate constructor(
     private val membershipGroupPolicyValidator: MembershipGroupPolicyValidator,
     @Reference(service = AllowedCertificatesReaderWriterService::class)
     private val allowedCertificatesReaderWriterService: AllowedCertificatesReaderWriterService,
+    @Reference(service = CordaAvroSerializationFactory::class)
+    serializationFactory: CordaAvroSerializationFactory,
```
Comment (yift-r3) [inline processors/db-processor/src/main/kotlin/net/corda/processors/db/internal/DBProcessorImpl.kt:121]: Do we need to add any of these to the `dependentComponents`?

File: processors/db-processor/src/main/kotlin/net/corda/processors/db/internal/reconcile/db/MemberInfoReconciler.kt [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+    private val keyValuePairListDeserializer: CordaAvroDeserializer<KeyValuePairList> by lazy {
+        cordaAvroSerializationFactory.createAvroDeserializer(
+            {
+                logger.error("Failed to deserialize key value pair list.")
+            },
+            KeyValuePairList::class.java
+        )
+    }
+    @VisibleForTesting
+    internal val reconcilerReadWriter = MemberInfoReconcilerReadWriter()
```
Comment (yift-r3) [inline processors/db-processor/src/main/kotlin/net/corda/processors/db/internal/reconcile/db/MemberInfoReconciler.kt:130]: You can add it to the constructor with a default construction...

File: processors/db-processor/src/main/kotlin/net/corda/processors/db/internal/reconcile/db/MemberInfoReconciler.kt [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+        ReconcilerWriter<String, PersistentMemberInfo>, ReconcilerReader<String, PersistentMemberInfo>,  Lifecycle {
+
+        override val lifecycleCoordinatorName = LifecycleCoordinatorName.forComponent<MemberInfoReconcilerReadWriter>()
+        private val coordinator = coordinatorFactory.createCoordinator(lifecycleCoordinatorName, ::handleEvent)
+        private val recordsMap = ConcurrentHashMap<String, PersistentMemberInfo>()
+
+        override fun getAllVersionedRecords(): Stream<VersionedRecord<String, PersistentMemberInfo>>? {
+            return recordsMap.entries.stream().mapNotNull {
+                val version = try {
+                    it.value.mgmContext.toSortedMap()[SERIAL]?.toInt()
```
Comment (yift-r3) [inline processors/db-processor/src/main/kotlin/net/corda/processors/db/internal/reconcile/db/MemberInfoReconciler.kt]: toSortedMap -> toMap (or just filter to make it faster)

File: processors/db-processor/src/main/kotlin/net/corda/processors/db/internal/reconcile/db/MemberInfoReconciler.kt [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+
+            override val keyClass = String::class.java
+            override val valueClass = PersistentMemberInfo::class.java
+
+            override fun onNext(
+                newRecord: Record<String, PersistentMemberInfo>,
+                oldValue: PersistentMemberInfo?,
+                currentData: Map<String, PersistentMemberInfo>
+            ) {
+                newRecord.value?.let { recordsMap[newRecord.key] = it } ?: recordsMap.remove(newRecord.key)
```
Comment (yift-r3) [inline processors/db-processor/src/main/kotlin/net/corda/processors/db/internal/reconcile/db/MemberInfoReconciler.kt:257]: Do we need to worry about memory consumption?

File: processors/db-processor/src/main/kotlin/net/corda/processors/db/internal/DBProcessorImpl.kt (hunk 4 above)
Comment (williamvigorr3) [reply processors/db-processor/src/main/kotlin/net/corda/processors/db/internal/DBProcessorImpl.kt:121]: `CordaAvroSerializationFactory`, `SubscriptionFactory,` and `PublisherFactory` don't have a lifecycle (so it should be Okay).

File: processors/db-processor/src/main/kotlin/net/corda/processors/db/internal/reconcile/db/MemberInfoReconciler.kt (hunk 5 above)
Comment (williamvigorr3) [reply processors/db-processor/src/main/kotlin/net/corda/processors/db/internal/reconcile/db/MemberInfoReconciler.kt:130]: `MemberInfoReconcilerReadWriter()` is an inner class, so I can't do that.

**Review by yift-r3 — DISMISSED** [review]

**Review by yift-r3 — DISMISSED** [review]
LGTM

**Review by yift-r3 — DISMISSED** [review]

**Review by seanbrereton — DISMISSED** [review]
LGTM (Gradle only)

File: gradle.properties [hunk 8]
```diff
@@ -45,8 +45,7 @@ commonsTextVersion = 1.10.0
 bouncycastleVersion=1.73
 # Corda API libs revision (change in 4th digit indicates a breaking change)
 # Change to 5.0.0.xx-SNAPSHOT to pick up maven local published copy
-#cordaApiVersion=5.1.0.xxx-SNAPSHOT
-cordaApiVersion=5.1.0.12-beta+
+cordaApiVersion=5.1.0.13-alpha-1692010685500
```
Comment (josephzunigadaly) [inline gradle.properties]: Reminder to revert before merge.

**Review by josephzunigadaly — DISMISSED** [review]

**Review by josephzunigadaly — DISMISSED** [review]

**Review by josephzunigadaly — DISMISSED** [review]

**Review by yift-r3 — DISMISSED** [review]

**Review by josephzunigadaly — APPROVED** [review]

**Review by yift-r3 — APPROVED** [review]

**Review by wzur-r3 — APPROVED** [review]
LGTM change for Gradle

**Pull request merged**